engine/src/main/java/org/terasology/engine/rendering/logic/SkeletonRenderer.java

```diff
@@ -366,30 +366,33 @@ public void renderOverlay() {
                         continue;
                     }
 
+                    // TODO: the position of the bone is de-coupled from the actual translation/scale
                     EntityRef boneParentEntity = skeletalMesh.boneEntities.get(parentBone.getName());
                     LocationComponent locCompA = boneEntity.getComponent(LocationComponent.class);
                     LocationComponent locCompB = boneParentEntity.getComponent(LocationComponent.class);
 
-                    // relative transform around the root skeletalMesh.scale and skeletalMesh.translate are relative to root
+                    // need to calculate the relative transformation from the entity to the start of the bone
                     locCompA.getRelativeTransform(relMat.identity(), entity);
+                    // entityTransform * (scale, translation) * relativeMat * [x,y,z,1]
                     result.set(entityTransform)
                             .mul(relFinal.identity()
                                     .scale(skeletalMesh.scale)
                                     .translate(skeletalMesh.translate)
-                                    .mul(relMat)) // apply relative transformation for the skeletalMesh scale and transform
-                            .transformPosition(currentPos.zero());
-                    meshData.position.put(currentPos);
+                                    .mul(relMat))
+                            .transformPosition(currentPos.zero()); // get the position of the start of the bone
+                    meshData.position.put(currentPos); // the start of the bone
 
-                    // relative transform around the root skeletalMesh.scale and skeletalMesh.translate are relative to root
+                    // need to calculate the relative transformation from the entity to the connecting bone
                     locCompB.getRelativeTransform(relMat.identity(), entity);
+                    // entityTransform * (scale, translation) * relativeMat * [x,y,z,1]
                     result.set(entityTransform)
                             .mul(relFinal
                                     .identity()
                                     .scale(skeletalMesh.scale)
                                     .translate(skeletalMesh.translate)
-                                    .mul(relMat))  // apply relative transformation for the skeletalMesh scale and transform
-                            .transformPosition(currentPos.zero());
-                    meshData.position.put(currentPos);
+                                    .mul(relMat))
+                            .transformPosition(currentPos.zero()); // get the position to the connecting bone
+                    meshData.position.put(currentPos); // the end of the bone
```

Comment:
Those two sections seem to do pretty much the same apart from the difference between `locCompA` and `locCompB` which according to the doc comments are the start of the bone vs the connecting bone.
Do you think it would make sense to extract this into a separate function and call it with the respective `locComp`?

Reply:
Would require passing a lot of parameters. Will be overhauled by @pollend soon-ish anyway, at which point we should ensure a comprehensible structure of the logic. Accepting it as is for now.